### PR TITLE
feat(frontend) Add PostgreSQL support

### LINKED ISSUE #3: Support for Postresql

### PR DESCRIPTION
**Pull request by Hywan**
Fix #3.

Please @cforce, can you test it on your database?

```sh
$ ./bin/database-to-plantuml -d 'pgsql:dbname=employees' -u … -s employees
```

I have a permission issue but queries are working, so I assume it works :-].

cc @jubianchi

### CODE REVIEW

**Comment by jubianchi** [comment]
I just tested this branch on one of my databases. Here is the output:

```
@startuml

!define table(x) class x << (T,#ffebf3) >>
hide methods
hide stereotypes
skinparam classFontColor #3b0018
skinparam classArrowColor #ff0066
skinparam classBorderColor #ff0066
skinparam classBackgroundColor ##f6f4ee
skinparam shadowing false

@enduml
```

File: src/Frontend/PgSQL/Database.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                    'FROM   information_schema.tables ' .
+                    'WHERE  table_catalog = :database_name ' .
+                    'AND    table_schema = :table_schema',
+                    [
+                        PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL
+                    ]
+                )
+                ->execute([
+                    'database_name' => $this->name,
+                    'table_schema'  => 'public'
```
Comment (jubianchi) [inline src/Frontend/PgSQL/Database.php]: This won't work in every case: using the `public` schema is kind of a bad practice. The user will likely have their own schema in the database. 

You have to provide a CLI argument to select the schema to reverse or provide an alternative syntax for PostgreSQL (and some other DB engine you would like to support like Oracle), something like `-s database_name.schema_name`

File: src/Frontend/PgSQL/Table.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                    'AND       c.table_name = :table_name ' .
+                    'AND       c.table_schema = :table_schema',
+                    [
+                        PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL
+                    ]
+                )
+                ->execute([
+                    'database_name' => $this->databaseName,
+                    'table_name'    => $this->name,
+                    'table_schema'  => 'public'
```
Comment (jubianchi) [inline src/Frontend/PgSQL/Table.php]: same here.

**Comment by Hywan** [comment]
@jubianchi I have the same result. I assume it is a permission issue because when running the queries by hand with `psql`, results are correct. But when running through PDO, the result is an empty array. Any idea why? The user is the same on both cases.

File: src/Frontend/PgSQL/Database.php (hunk 1 above)
Comment (Hywan) [reply src/Frontend/PgSQL/Database.php]: Here we go: https://github.com/Hywan/Database-to-PlantUML/pull/4/commits/423c2c314505c5c5cc314af313568bf9f5c0a1ae. Thoughts?

**Comment by jubianchi** [comment]
@Hywan I don't think it's a permission issue because we don't have any exception.

I think there is something wrong with the database layer. If I apply this patch:

```diff
diff --git i/src/Frontend/PgSQL/Database.php w/src/Frontend/PgSQL/Database.php
index 3cc5a96..cff6f2b 100644
--- i/src/Frontend/PgSQL/Database.php
+++ w/src/Frontend/PgSQL/Database.php
@@ -12,33 +12,18 @@ class Database extends Frontend\Database
 {
     public function tables(): iterable
     {
-        $tables =
-            $this->getDatabaseConnection()
-                ->prepare(
-                    'SELECT table_catalog AS databaseName, ' .
-                    '       table_name AS name ' .
-                    'FROM   information_schema.tables ' .
-                    'WHERE  table_catalog = :database_name ' .
-                    'AND    table_schema = :table_schema',
-                    [
-                        PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL
-                    ]
-                )
-                ->execute([
-                    'database_name' => $this->name,
-                    'table_schema'  => 'public'
-                ]);
+        $pdo = new PDO('pgsql:host=localhost port=2254 dbname=ods_instance0', 'postgres');
+        $stmt = $pdo->prepare('SELECT table_catalog AS databaseName, ' .
+            '       table_name AS name ' .
+            'FROM   information_schema.tables ' .
+            'WHERE  table_catalog = :database_name ' .
+            'AND    table_schema = :table_schema');
+        $stmt->setFetchMode(PDO::FETCH_CLASS, Table::class, [$this->getDatabaseConnection()]);
+        $stmt->execute([
+            'database_name' => $this->name,
+            'table_schema'  => 'authentication'
+        ]);
 
-        $tables->setFetchingStyle(
-            DalStatement::FROM_START,
-            DalStatement::FORWARD,
-            DalStatement::AS_CLASS,
-            Table::class,
-            [
-                $this->getDatabaseConnection()
-            ]
-        );
-
-        yield from $tables;
+        return new \ArrayIterator($stmt->fetchAll());
     }
 }

```

I get the table list and a correct PlantUML output:

```
@startuml

!define table(x) class x << (T,#ffebf3) >>
hide methods
hide stereotypes
skinparam classFontColor #3b0018
skinparam classArrowColor #ff0066
skinparam classBorderColor #ff0066
skinparam classBackgroundColor ##f6f4ee
skinparam shadowing false

table(foo) {
}


table(bar) {
}

@enduml
```

**Comment by jubianchi** [comment]
Ok I think I found something: you are using the `` and it does not seem to work well with `pdo_pgsql`. I replaced it with `CURSOR_FWDONLY` and its seems to be better.

I still can't get the column list, I'm trying to find why.

**Comment by jubianchi** [comment]
@Hywan the `$databaseName` property on `Hywan\DatabaseToPlantUML\Frontend\Table` seems to be `NULL`. This is why we can't get the column list.

**Comment by Hywan** [comment]
Because alias in PgSQL are put in lowercase. `table_catalog AS databaseName` is reduced to `table_catalog AS databasename`:

```diff
- databaseName
+ databasename
```

**Comment by Hywan** [comment]
We need to write it this way: `table_catalog AS "databaseName"` and it works. Will commit everything.

**Comment by Hywan** [comment]
@jubianchi Can you retry now? It works for me.

File: src/Frontend/PgSQL/Table.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+use Hywan\DatabaseToPlantUML\Frontend;
+use PDO;
+
+class Table extends Frontend\Table
+{
+    const PRIMARY = '/_pkey$/';
+
+    public function columns(): iterable
+    {
+        list($databaseName, $tableSchema) = array_merge(explode('.', $this->databaseName), [1 => 'public']);
```
Comment (Hywan) [inline src/Frontend/PgSQL/Table.php]: `$this->databaseName` will never contain a dot here. Need to be fixed.

**Comment by Hywan** [comment]
It works, but it's not correct :-)

output

**Comment by jubianchi** [comment]
@Hywan yes, there is a problem with joins.

**Comment by Hywan** [comment]
output

Much better :-)

**Pull request merged**